enterprise/tpa_pipeline.py

```diff
@@ -0,0 +1,153 @@
+"""
+Module provides elements to be used in third-party auth pipeline
+"""
+from django.core.urlresolvers import reverse
+from django.shortcuts import redirect
+from django.utils.translation import ugettext as _
+
+try:
+    from third_party_auth.provider import Registry
+except ImportError:
+    Registry = None
+
+from enterprise.models import EnterpriseCustomer, UserDataSharingConsentAudit
+from enterprise.utils import NotConnectedToEdX
+
+from social.pipeline import partial
+
+class MultipleCustomersFound(Exception):
+    """
+    Raise this exception if there are multiple customers tied to an IdP.
+    """
+
+
+def get_enterprise_customer_for_request(request):
+    """
+    Get the EnterpriseCustomer associated with a particular request.
+    """
+    if Registry is None:
+        raise NotConnectedToEdX(
+            _("This package must be installed in an EdX environment to look up third-party auth providers.")
+        )
+    pipeline = request.session.get('partial_pipeline')
+    provider = Registry.get_from_pipeline(pipeline)
+    return get_enterprise_customer_for_sso(provider)
+
+
+def get_enterprise_customer_for_running_pipeline(pipeline):
+    """
+    Get the EnterpriseCustomer associated with a running pipeline
+    """
+    if Registry is None:
+        raise NotConnectedToEdX(
+            _("This package must be installed in an EdX environment to look up third-party auth providers.")
+        )
+
+    provider = Registry.get_from_pipeline(pipeline)
+    return get_enterprise_customer_for_sso(provider)
+
+
+def get_enterprise_customer_for_sso(provider):
+    """
+    Get the EnterpriseCustomer object tied to an identity provider.
+    """
+    if provider is None:
+        return None
+    try:
+        return EnterpriseCustomer.objects.get(identity_provider=provider.idp_slug)
```

Comment:
I assume this isn't your code, but we have an opportunity here to fix this to make it more general: The `identity_provider` field of Enterprise Customer should reference a complete [`provider_id`](https://github.com/edx/edx-platform/blob/f806f3a/common/djangoapps/third_party_auth/models.py#L158-L162) such as "`saml-mycorp`" and not just a SAML `provider_slug` like "`mycorp`". That way, we can support non-SAML providers in the future if we need to (OAuth2, Active Directory, or other types). We should not tie ourselves unnecessarily to SAML if the code can be fully generic.

This line would become
```
EnterpriseCustomer.objects.get(identity_provider=provider.provider_id)
```